addons/binding/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/internal/ssl/SSLContextBuilder.java

```diff
@@ -0,0 +1,215 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.energenie.internal.ssl;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.text.MessageFormat;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A helper class for building a {@link SSLContext}
+ *
+ * @author Svilen Valkanov
+ */
+
+public class SSLContextBuilder {
+
+    private static final String X_509_CERTIFICATE = "X.509";
+    private static final String JKS_KEYSTORE_TYPE = "JKS";
+    private static final String SSL_PROTOCOL = "SSL";
+
+    private final Logger logger = LoggerFactory.getLogger(SSLContextBuilder.class);
+
+    private BundleContext bundleContext;
+    private TrustManager[] trustedManagers;
+    private KeyManager[] keyManagers;
+    private SecureRandom randomGenerator;
+
+    private SSLContextBuilder(BundleContext bundleContext) {
+        this.bundleContext = bundleContext;
+    }
+
+    /**
+     * Creates a SSLContextBuilder that can load key stores and certificates from a bundle
+     *
+     * @param bundleContext - used to load files from a bundle
+     * @return SSLContextBuilder
+     */
+
+    public static SSLContextBuilder create(BundleContext bundleContext) {
+        return new SSLContextBuilder(bundleContext);
+    }
+
+    /**
+     * Initializes the {@link SSLContext} using the appended {@link TrustManager}s and {@link KeyManager}s and
+     * {@link SecureRandom} generator
+     *
+     * @return SSLContext or null if no {@link TrustManager}s or {@link KeyManager}s are set
+     */
+    public SSLContext build() {
+        try {
+            if (trustedManagers != null || keyManagers != null) {
+                SSLContext sslContext = SSLContext.getInstance(SSL_PROTOCOL);
+                sslContext.init(keyManagers, trustedManagers, randomGenerator);
+                return sslContext;
+            }
+        } catch (KeyManagementException e) {
+            logger.error("A KeyManagementException occurred", e);
+        } catch (NoSuchAlgorithmException e) {
+            logger.error("A NoSuchAlgorithmException occurred", e);
+        }
+        return null;
```

Comment:
This part is way better than it was but I'm still curious what happens when you return `null` and call `setSslContext` with `null`. If that is accepted I think you are not able to make any connection to an SSL site then or the whole exercise was not needed and we already had a proper truststore before doing this.

I think that this line should throw an error because or the builder was not properly initialized or we had an exception.